### PR TITLE
O317 - Add header for flash messages

### PR DESCRIPTION
**Pull request by razorcd**
https://osraav.atlassian.net/browse/OSRA-317

Created _flash.html.erb partial to display success and error flash messages.
+ rspec tests

To show a flash masseges anywhere in the rails views just use:

flash[:success] = "Partner saved"       or
flash[:error] = "Partner failed to save"       or use the form validation errors instead

### CODE REVIEW

File: app/helpers/_flashes_helper.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module FlashesHelper
+  
+  def flash_htmlclass_for type
+    case type.to_sym 
+      when :success 
+        type_class = 'alert alert-success'
```
Comment (theB264) [inline app/helpers/_flashes_helper.rb]: Do we need `type_class=` here?

**Comment by razorcd** [comment]
no we don't, I'll fix that. Thanks!

File: app/helpers/_flashes_helper.rb (hunk 1 above)
Comment (razorcd) [inline app/helpers/_flashes_helper.rb]: no we don't, I'll fix that. Thanks!

**Comment by razorcd** [comment]
Looks like this:

error_messages

**Comment by PurityControl** [comment]
:+1:

File: app/helpers/_flashes_helper.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+  def flash_htmlclass_for type
+    case type.to_sym 
+      when :success 
+        'alert alert-success'
+      when :error
+        'alert alert-danger'
+    end
+  end
+
+end
```
Comment (PurityControl) [inline app/helpers/_flashes_helper.rb]: newline?

File: app/views/layouts/_flashes.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<% unless flash.empty? %>
+  <div class="flashes">
+    <% flash.each do |type, message| %>
+      <div class="<%= flash_htmlclass_for type %>"><%= message.to_s %></div>
+    <% end %>
+  </div>
+<% end %>
```
Comment (PurityControl) [inline app/views/layouts/_flashes.html.erb]: newline?

Comment (razorcd) [inline app/views/layouts/_flashes.html.erb]: what do you mean?

Comment (theB264) [inline app/views/layouts/_flashes.html.erb]: Add another blank line to the file

Comment (razorcd) [inline app/views/layouts/_flashes.html.erb]: ah, last line should be empty, got it. Thanks!

Comment (theB264) [inline app/views/layouts/_flashes.html.erb]: Many editors can be set up to automatically ensure a trailing newline.

Comment (NikitaAvvakumov) [inline app/views/layouts/_flashes.html.erb]: That's it, I'm putting style guide back on the agenda!

Comment (theB264) [inline app/views/layouts/_flashes.html.erb]: :+1:

**Comment by razorcd** [comment]
PR updated
I added the newline at eof. Github does not show this in view mode, it only shows in edit mode. Don't know why.

**Comment by theB264** [comment]
:shipit:

File: app/views/layouts/_flashes.html.erb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<% unless flash.empty? %>
+  <div class="flashes">
+    <% flash.each do |type, message| %>
+      <div class="<%= flash_htmlclass_for type %>"><%= message.to_s %></div>
```
Comment (PurityControl) [inline app/views/layouts/_flashes.html.erb]: sorry to ask this so late in the game but
1. what happens (in terms of the css class) when anything other than type :success or :error is called. 
  - Do we have just an empty class attribute?
  - Is this valid html?
  - Do we want to have a default we return such as class="flash", and if we do do we want to add this class to the other cases too.

Comment (razorcd) [inline app/views/layouts/_flashes.html.erb]: I considered that the flash messages can be only `:success` or `:error` as we agreed in the scrum.

**Comment by razorcd** [comment]
pr updated. Flash messages only accept   `:success`   and   `:error`   keys. For only any other flash keys it will not show the `<div class="flashes">` at all.

File: app/helpers/_flashes_helper.rb [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  end
+
+  def each_accepted_flash
+    flash.each do |fl|
+      yield fl if flash_htmlclass_for fl[0]
+    end
+  end
+
+  def any_accepted_flash?
+    flash.any? {|k,v| flash_htmlclass_for k }
```
Comment (NikitaAvvakumov) [inline app/helpers/_flashes_helper.rb]: Minor point: prefer to use underscore for unused block parameters. i.e. `{ |k, _| flash_htmlclass_for k }`.
Even more minor point: prefer consistent padding (or not) of curly braces with spaces.

File: app/helpers/_flashes_helper.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+module FlashesHelper
+
+  ACCEPTED_FLASH_HTML_CLASSES = {
+    success: 'alert alert-success',
+    error: 'alert alert-danger'
```
Comment (NikitaAvvakumov) [inline app/helpers/_flashes_helper.rb]: Is there any interest in adding the two other Bootstrap alert styles at this point?
Do we want alerts to be dismissible?

File: app/helpers/_flashes_helper.rb (hunk 5 above)
Comment (razorcd) [inline app/helpers/_flashes_helper.rb]: _ for unused vars, I like that!

File: app/helpers/_flashes_helper.rb (hunk 6 above)
Comment (razorcd) [inline app/helpers/_flashes_helper.rb]: We decided at the scrum from 1st Jan that we will only use these 2. (green and red)
Not sure what to say about dismissible alerts. They look cool.

File: app/views/layouts/_flashes.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+    <% each_accepted_flash do |type, message| %>
+      <div class="<%= flash_htmlclass_for type %>">
+        <%= button_to "#", :class => "close", "data-dismiss" => "alert", :aria_label => "Close" do %>
+          <span aria-hidden="true">&times;</span>
+        <% end %>
+        <%= message.to_s %>
+      </div>
+    <% end %>
+  </div>
+<% end %>
```
Comment (razorcd) [inline app/views/layouts/_flashes.html.erb]: Flash dismiss button has been added.

**Comment by razorcd** [comment]
PR updated.
Looks like this:
error_messages3

**Comment by theB264** [comment]
:+1:

**Pull request merged**